atlasdb-commons/src/main/java/com/palantir/util/IndexEncodingUtils.java

```diff
@@ -0,0 +1,121 @@
+            Set<K> keys, Map<K, V> keyToValue, Function<V, R> valueMapper, Function<K, byte[]> byteMapper) {
+        List<K> keyList = new ArrayList<>(keys);
+        // A linked hash map will give a minor improvement when iterating during serialization
+        Map<Integer, R> indexToValue = Maps.newLinkedHashMapWithExpectedSize(keyToValue.size());
+        // We are explicitly using a primitive for-loop (no streaming) to squeeze out a bit of performance
+        // (same in decode)
+        for (int i = 0; i < keyList.size(); i++) {
+            V value = keyToValue.get(keyList.get(i));
+            if (value != null) {
+                indexToValue.put(i, valueMapper.apply(value));
+            }
+        }
+        Preconditions.checkArgument(
+                indexToValue.size() >= keyToValue.size(), "Value map contains keys that are not in the key list");
+        return IndexEncodingResult.of(keyList, indexToValue, computeCheckSum(keyList, byteMapper));
+    }
+
+    /**
+     * Given an ordered list of keys and a map of indices to values, returns a map of keys to values.
+     * The values are determined by indexing into the list of keys.
+     * Values of the returned map are created using the provided mapping function.
+     * Also computes a checksum of the ordered keys using a provided function and compares it to the provided checksum.
+     *
+     * @param indexEncodingResult the output of {@link IndexEncodingUtils#encode}, i.e. the ordered list of keys,
+     * a map of indices to values, and a checksum of the ordered key list used to compute it. Every index MUST be
+     * a valid index into the list of keys.
+     * @param valueMapper a mapping function to run before putting a value into the result map
+     * @param byteMapper a function that extracts bytes from a key object, which are used to compute the checksum
+     * @throws SafeIllegalArgumentException if the provided checksum does not match the checksum of the ordered keys
+     */
+    public static <K, V, R> Map<K, R> decode(
+            IndexEncodingResult<K, V> indexEncodingResult, Function<V, R> valueMapper, Function<K, byte[]> byteMapper) {
+        Preconditions.checkArgument(
+                computeCheckSum(indexEncodingResult.keyList(), byteMapper)
+                        == indexEncodingResult.keyListCrc32Checksum(),
+                "Key list integrity check failed");
+
+        Map<Integer, V> indexToValue = indexEncodingResult.indexToValue();
+        Map<K, R> keyToValue = Maps.newHashMapWithExpectedSize(indexToValue.size());
+        for (Map.Entry<Integer, V> entry : indexToValue.entrySet()) {
+            keyToValue.put(indexEncodingResult.keyList().get(entry.getKey()), valueMapper.apply(entry.getValue()));
+        }
+        return keyToValue;
+    }
+
+    private static <K> long computeCheckSum(List<K> keyList, Function<K, byte[]> byteMapper) {
+        CRC32 orderedKeyChecksum = new CRC32();
```

Comment:
I was thinking what would happen if we change our check sum algorithm/approach in the future. Maybe we should use a more general `checksum` field in the result and also return the type of checksum algorithm is used. This would allow for more seamless upgrades of the checksum computation without having to disable integrity checks.

Reply:
My general opinion is that the checksum should be recorded as a compound object recording the method used and the value, to facilitate changing it easily, especially since we're not relying on this for security. That said, what's the wire overhead of doing the integrity check and method?

Reply:
I have added a `KeyListCheckSum` compound type to record the type and value of the checksum. I am not sure what you are getting to with your last question. The overhead of the integrity check + method should be negligible in comparison to the data (lock descriptors) and metadata.

Reply:
Following up: We could omit the integrity check for single-element lists (as those can trivially not be reordered) and make it optional when transmitting over the wire. This would also enable easier backwards-compatibility if we choose to remove checksums in the future.